### ISSUE TITLE
Training an existing Spacy Ner pipeline forgets previous examples

### ISSUE THREAD

**Issue by suityou01** [explosion/spaCy#4994]
I am creating a new model for Named entity recognition.
I have a training data set that looks like :

    [[
    "world leading global energy trading company has an exciting 
    opportunity for a senior csharp application platform developer to 
    join its systematic trading division developing new trading 
    applications tools and solutions for its successful front office 
    trading team", 
    {"entities": [[80, 86, "RNK"]]}
    ]]

So I run the following function to train the model.

    @plac.annotations(
    model=("Model name. Defaults to blank 'en' model.", "option", "m", 
    str),
    new_model_name=("New model name for model meta.", "option", "nm", 
    str),
    output_dir=("Optional output directory", "option", "o", Path),
    n_iter=("Number of training iterations", "option", "n", int), 
    entity=("Name of the entity to be trained", "option", "e", str),
    label=("The label to be given to the trained entity", "option", "l", 
    str),
    )
    def main(model = None, new_model_name=("ner%s" % str(datetime.now)), 
    output_dir= None, n_iter=20, entity=None, label=None):
    
    if entity is None or label is None:
        log.info("Entity and Label must both be supplied")
        log.info("Bailing out as nothing to do ...... :-(")
        return

    log.info("Fetching training data for entity [%s] to be trained with 
    label [%s]" % (entity, label))

    log.info("Training data retrieved and the first row is : ")
    log.info(TRAIN_DATA[0])
    log.info("There are %d rows to be trained" % len(TRAIN_DATA))
    
    if model is not None:
        nlp = spacy.load(output_dir)  # load existing spaCy model
        print("Loaded model '%s'" % model)
    else:
        nlp = spacy.blank("en")  # create blank Language class
        print("Created blank 'en' model")

    if "ner" not in nlp.pipe_names:
        log.info("ner not in pipe names, adding it in now ....")
        ner = nlp.create_pipe("ner")
        nlp.add_pipe(ner)
    # otherwise, get it, so we can add labels to it
    else:
        log.info("retrieving previous ner pipe now ....")
        ner = nlp.get_pipe("ner")

    # add labels
    for _, annotations in TRAIN_DATA:
         for ent in annotations.get('entities'):
            ner.add_label(ent[2])

    move_names = list(ner.move_names)

    # get names of other pipes to disable them during training
    pipe_exceptions = ["ner", "trf_wordpiecer", "trf_tok2vec"]
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe not in 
    pipe_exceptions]
    with nlp.disable_pipes(*other_pipes):  # only train NER
        # reset and initialize the weights randomly – but only if we're
        # training a new model
        if model is None:
            optimizer=nlp.begin_training()
        else:
            optimizer=nlp.resume_training()
        for itn in range(n_iter):
            random.shuffle(TRAIN_DATA)
            losses = {}
            # batch up the examples using spaCy's minibatch
            batches = minibatch(TRAIN_DATA, size=compounding(4.0, 32.0, 
   1.001))
            for batch in batches:
                texts, annotations = zip(*batch)
                nlp.update(
                    texts,  # batch of texts
                    annotations,  # batch of annotations
                    drop=0.5,  # dropout - make it harder to memorise 
    data
                    sgd=optimizer,
                    losses=losses,
                )
            print("Losses", losses)

    # test the trained model
    for text, _ in TRAIN_DATA:
        doc = nlp(text)
        print("Entities", [(ent.text, ent.label_) for ent in doc.ents])
        print("Tokens", [(t.text, t.ent_type_, t.ent_iob) for t in doc])

    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()
        nlp.meta["name"] = new_model_name  # rename model
        nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the saved model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        # Check the classes have loaded back consistently
        assert nlp2.get_pipe("ner").move_names == move_names
        
    return

    if __name__ == "__main__":
        log = logging.getLogger(__name__)
        log.setLevel(logging.DEBUG)

        consoleHandler = logging.StreamHandler()
        consoleHandler.setLevel(logging.DEBUG)

        log.addHandler(consoleHandler)
        plac.call(main)

This works well and I am able to then test the model successfully using the following function

    @plac.annotations(
    model_dir=("Optional output directory", "option", "o", Path),
    test_text=("The test text to be used to test the model","option", 
    "t", str),
    entity=("Name of the entity to be trained", "option", "e", str),
    label=("The label to be given to the trained entity", "option", "l", 
    str),
    )
    def main(model_dir= None,test_text=None, entity=None, label=None):
    
    if entity is None or label is None:
        log.info("Entity and Label must both be supplied")
        log.info("Bailing out as nothing to do ...... :-(")
        return
    
    if test_text is None:
        test_text = ("Using default test string which is not optimal to 
    look for %s" % entity)
    
    nlp = spacy.load(model_dir)
    log.info("Loaded model %s" % nlp.meta["name"])
    log.info("Testing the string %s" % test_text)

    ner = nlp.get_pipe("ner")
    for label in ner.labels:
        log.info("NER Label : %s found in model" % label)

    doc = nlp(test_text)
    print("Entities", [(ent.text, ent.label_) for ent in doc.ents])
    print("Tokens", [(t.text, t.ent_type_, t.ent_iob) for t in doc])

    if __name__ == "__main__":
    log = logging.getLogger(__name__)
    log.setLevel(logging.DEBUG)

    consoleHandler = logging.StreamHandler()
    consoleHandler.setLevel(logging.DEBUG)

    log.addHandler(consoleHandler)
    plac.call(main)

However, if I run the code again to train against a new label, say "SKILL" and provide new training data, it successfully loads the old model, retrieves the Ner pipeline and sets the optimizer to resume, but when I test it again it has forgotten all of it's training about the RNK label.

I assume that resume would somehow pick up with a previous model state and retain previously learned annotations. It certainly retains the NER label.

Why is this happening?

Thinking this was perhaps related to the catastrophic forgetting problem I created one big set of training data with examples from both categories,so that is :

    'TOKEN'  >>> 'NER Annotation'
    -----------------------------
    'senior' >>> 'RNK'
    'csharp' >>> 'SKILL'
    'sql'    >>> 'SKILL'

These are the losses

    Losses {'ner': 721.8737016180717}
    Losses {'ner': 5.999976082008388}
    Losses {'ner': 5.970323057037423}
    Losses {'ner': 5.996330579093365}
    Losses {'ner': 6.028536462566022}
    Losses {'ner': 12.043830573641666}
    Losses {'ner': 10.001897952651317}
    Losses {'ner': 6.016950026187274}
    Losses {'ner': 6.624311646328313} 
    Losses {'ner': 10.602919933949224}
    Losses {'ner': 6.1062697231067995}
    Losses {'ner': 8.792055106010444}
    Losses {'ner': 13.302123281119345}
    Losses {'ner': 6.068028368915684}
    Losses {'ner': 8.026694430880903}
    Losses {'ner': 8.961434860193798}
    Losses {'ner': 6.02721516249698}
    Losses {'ner': 9.714660156853073}
    Losses {'ner': 4.108544494319015}
    Losses {'ner': 6.023105974059858}
    Losses {'ner': 7.357760648981275}
    Losses {'ner': 6.295292869532734}
    Losses {'ner': 3.8088561052881995}
    Losses {'ner': 6.059279332644757}
    Losses {'ner': 7.024559462190113}
    Losses {'ner': 4.784358718788942}
    Losses {'ner': 5.935101364429172}
    Losses {'ner': 4.027772727507415}
    Losses {'ner': 2.1748163004265884}
    Losses {'ner': 5.993975825343896}

I have included my training data on pastebin, which can be found here :

https://pastebin.com/JmQv7HRS

**Comment by adrianeboyd** [comment]
I think your model doesn't have any annotations to learn from because there are alignment problems with the character offsets in the training data.

The character offsets for the string spans should look like string slices (the end offset is exclusive), so instead of:

```
['senior developerdesigner', {'entities': [[0, 5, 'RNK']]}]
```

you'd want:

```
['senior developerdesigner', {'entities': [[0, 6, 'RNK']]}]
```

While training, spacy silently discards entity spans that don't line up with token boundaries (this should definitely be improved so it's more obvious to users!), so you want to check your annotation carefully to make sure it's aligned like you expect.

I'd recommend checking with a script like this to see whether you end up with `None` spans where it doesn't know how to align the annotation:

```
from spacy.lang.en import English # use the language you're training for

TRAIN_DATA = [
    ("Who is Shaka Khan?", {"entities": [(1, 5, "PERSON")]}), # MISALIGNED!
    ("I like London and Berlin.", {"entities": [(7, 13, "LOC"), (18, 24, "LOC")]}),
]

nlp = English()
for text, annot in TRAIN_DATA:
    doc = nlp(text)
    print(text, "\t", [doc.char_span(start_idx, end_idx, label=label) for start_idx, end_idx, label in annot["entities"]])
```

Output:

```
Who is Shaka Khan? 	 [None]
I like London and Berlin. 	 [London, Berlin]
```

Cases where you see `None` are ones that are misaligned with the default English tokenization. Depending on your task, you might want to modify the offsets or you might want to modify the tokenizer.
